### PR TITLE
[bitnami/prometheus-operator] new chart

### PR DESCRIPTION
**Pull request by sameersbn**
**Description of the change**

New chart for deploying Prometheus Operator to the cluster

**Benefits**

Prometheus Operator provides easy monitoring definitions for Kubernetes services and deployment and management of Prometheus instances.

**Possible drawbacks**

N/A

**Applicable issues**

N/A

**Additional information**

N/A

**Checklist** 
- [x] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/).
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[bitnami/chart]`)
- [x] If the chart contains a `values-production.yaml` apart from `values.yaml`, ensure that you implement the changes in both files

### CODE REVIEW

File: bitnami/prometheus-operator/README.md [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+| `operator.readinessProbe.successThreshold`            | Minimum consecutive successes for the probe                                          | `1`                                                                  |
+| `operator.logLevel`                                   | Log Level                                                                            | `info`                                                               |
+| `operator.logFormat`                                  | Log Format                                                                           | `logfmt`                                                             |
+| `operator.kubeletService.enabled`                     | Whether to maintain a service for scraping kubelets                                  | `true`                                                               |
+| `operator.kubeletService.namepsace`                   | Namespace to deploy the kubelet service                                              | `kube-system`                                                        |
```
Comment (javsalgar) [inline bitnami/prometheus-operator/README.md]: namepsace -> namespace

File: bitnami/prometheus-operator/README.md [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+| `prometheus.remoteWrite`                      | The remote_write spec configuration for Prometheus                                   | `[]`                                                                   |
+| `prometheus.storageSpec`                      | Prometheus StorageSpec for persistent data                                           | `{}`                                                                   |
+| `prometheus.priorityClassName`                | Priority class assigned to the Pods                                                  | ``                                                                     |
+| `prometheus.containers`                       | Containers allows injecting additional containers                                    | `[]`                                                                   |
+| `prometheus.additionalScrapeConfigsExternal`  | Enable additional scrape configs that are managed externally to this chart           | `false`                                                                |
```
Comment (javsalgar) [inline bitnami/prometheus-operator/README.md]: Missing value `prometheus.podSecurityPolicy.allowedCapabilities`

File: bitnami/prometheus-operator/README.md [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+
+## Introduction
+
+This chart bootstraps [Prometheus Operator](https://github.com/bitnami/bitnami-docker-prometheus-operator) on [Kubernetes](http://kubernetes.io) using the [Helm](https://helm.sh) package manager.
+
+Bitnami charts can be used with [Kubeapps](https://kubeapps.com/) for deployment and management of Helm Charts in clusters.
+
+## Prerequisites
+
+- Kubernetes 1.10+
```
Comment (javsalgar) [inline bitnami/prometheus-operator/README.md]: Change it to:

  - Kubernetes 1.12+
  - Helm 2.11+ or Helm 3.0-beta3+

File: bitnami/prometheus-operator/templates/alertmanager/ingress.yaml [hunk 4]
```diff
@@ -0,0 +1,29 @@
+{{- if and .Values.alertmanager.enabled .Values.alertmanager.ingress.enabled }}
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+  name: {{ template "prometheus-operator.alertmanager.fullname" . }}
+  labels: {{ include "prometheus-operator.alertmanager.labels" . | nindent 4 }}
```
Comment (javsalgar) [inline bitnami/prometheus-operator/templates/alertmanager/ingress.yaml]: When using `nindent` make sure you add `{{- ` so the rendered yaml does not contain trailing whitespaces

File: bitnami/prometheus-operator/templates/alertmanager/service.yaml [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  {{- end }}
+  ports:
+  - name: http
+    port: {{ .Values.alertmanager.service.port }}
+    targetPort: 9093
+    {{- if and .Values.alertmanager.service.nodePort (or (eq .Values.alertmanager.service.type "NodePort") (eq .Values.alertmanager.service.type "LoadBalancer")) }}
+    nodePort: {{ .Values.alertmanager.service.nodePort }}
+    {{- end }}
+  selector:
+    app: alertmanager
```
Comment (javsalgar) [inline bitnami/prometheus-operator/templates/alertmanager/service.yaml]: Instead of `app`, you could use `app.kubernetes.io/name`

Comment (javsalgar) [reply bitnami/prometheus-operator/templates/alertmanager/service.yaml]: Wait, I see that you use it. Not sure about the role of `app: alertmanager` here.

File: bitnami/prometheus-operator/values.yaml [hunk 6]
```diff
@@ -0,0 +1,726 @@
...
+operator:
+  enabled: true
+
+  ## Bitnami Prometheus Operator image version
+  ## ref: https://hub.docker.com/r/bitnami/prometheus-operator/tags/
+  ##
+  image:
+    registry: docker.io
+    repository: bitnami/prometheus-operator
+    tag: 0.33.0
```
Comment (javsalgar) [inline bitnami/prometheus-operator/values.yaml]: Make sure you are not using rolling tags here

File: bitnami/prometheus-operator/templates/alertmanager/alertmanager.yaml [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+  {{- end }}
+  listenLocal: {{ .Values.alertmanager.listenLocal }}
+  {{- if and .Values.alertmanager.ingress.enabled .Values.alertmanager.ingress.hosts }}
+  externalUrl: {{ if .Values.alertmanager.ingress.tls }}https{{else}}http{{ end }}://{{ (index .Values.alertmanager.ingress.hosts 0).name }}/
+  {{- else }}
+  externalUrl: http://{{ template "prometheus-operator.alertmanager.fullname" . }}.{{ .Release.Namespace }}:{{ .Values.alertmanager.service.port }}
+  {{- end }}
+  paused: {{ .Values.alertmanager.paused }}
+  logFormat: {{ .Values.alertmanager.logFormat }}
+  logLevel:  {{ .Values.alertmanager.logLevel }}
```
Comment (javsalgar) [inline bitnami/prometheus-operator/templates/alertmanager/alertmanager.yaml]: Extra space

File: bitnami/prometheus-operator/templates/alertmanager/psp-clusterrole.yaml [hunk 8]
```diff
@@ -0,0 +1,13 @@
+{{- if and .Values.alertmanager.enabled .Values.rbac.create .Values.rbac.pspEnabled }}
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: {{ template "prometheus-operator.alertmanager.fullname" . }}-psp
+  labels: {{ include "prometheus-operator.alertmanager.labels" . | nindent 4 }}
+rules:
+- apiGroups: ['extensions']
+  resources: ['podsecuritypolicies']
+  verbs:     ['use']
```
Comment (javsalgar) [inline bitnami/prometheus-operator/templates/alertmanager/psp-clusterrole.yaml]: Extra spaces

File: bitnami/prometheus-operator/templates/prometheus-operator/psp-clusterrole.yaml [hunk 9]
```diff
@@ -0,0 +1,13 @@
+{{- if and .Values.operator.enabled .Values.rbac.create .Values.rbac.pspEnabled }}
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: {{ template "prometheus-operator.operator.fullname" . }}-psp
+  labels: {{ include "prometheus-operator.operator.labels" . | nindent 4 }}
+rules:
+- apiGroups: ['extensions']
+  resources: ['podsecuritypolicies']
+  verbs:     ['use']
```
Comment (javsalgar) [inline bitnami/prometheus-operator/templates/prometheus-operator/psp-clusterrole.yaml]: Extra spaces

File: bitnami/prometheus-operator/templates/prometheus/prometheus.yaml [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+  {{- else if .Values.prometheus.replicaExternalLabelName }}
+  replicaExternalLabelName: "{{ .Values.prometheus.replicaExternalLabelName }}"
+  {{- end }}
+  {{- if and .Values.prometheus.ingress.enabled .Values.prometheus.ingress.hosts }}
+  externalUrl: {{ if .Values.prometheus.ingress.tls }}https{{else}}http{{ end }}://{{ (index .Values.prometheus.ingress.hosts 0).name }}/
+  {{- else }}
+  externalUrl: http://{{ template "prometheus-operator.prometheus.fullname" . }}.{{ .Release.Namespace }}:9090
+  {{- end }}
+  paused: {{ .Values.prometheus.paused }}
+  logLevel:  {{ .Values.prometheus.logLevel }}
```
Comment (javsalgar) [inline bitnami/prometheus-operator/templates/prometheus/prometheus.yaml]: Extra spaces

File: bitnami/prometheus-operator/templates/prometheus/psp-clusterrole.yaml [hunk 11]
```diff
@@ -0,0 +1,13 @@
+{{- if and .Values.prometheus.enabled .Values.rbac.create .Values.rbac.pspEnabled }}
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: {{ template "prometheus-operator.prometheus.fullname" . }}-psp
+  labels: {{ include "prometheus-operator.prometheus.labels" . | nindent 4 }}
+rules:
+- apiGroups: ['extensions']
+  resources: ['podsecuritypolicies']
+  verbs:     ['use']
```
Comment (javsalgar) [inline bitnami/prometheus-operator/templates/prometheus/psp-clusterrole.yaml]: Extra spaces

File: bitnami/prometheus-operator/templates/prometheus/prometheus.yaml [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+  logFormat:  {{ .Values.prometheus.logFormat }}
```
Comment (javsalgar) [inline bitnami/prometheus-operator/templates/prometheus/prometheus.yaml]: Extra spaces

File: bitnami/prometheus-operator/templates/alertmanager/service.yaml (hunk 5 above)
Comment (sameersbn) [reply bitnami/prometheus-operator/templates/alertmanager/service.yaml]: These labels are added by the `prometheus-operator` component. We have to use whatever labels they've defined in the code

ref: https://github.com/coreos/prometheus-operator/blob/master/pkg/alertmanager/statefulset.go#L195-L197

File: bitnami/prometheus-operator/README.md [hunk 13]
```diff
@@ -0,0 +1,304 @@
...
+
+### [Rolling VS Immutable tags](https://docs.bitnami.com/containers/how-to/understand-rolling-tags-containers/)
+
+It is strongly recommended to use immutable tags in a production environment. This ensures your deployment does not change automatically if the same tag is updated with a different image.
+
+Bitnami will release a new chart updating its containers if a new version of the main container, significant changes, or critical vulnerabilities exist.
+
+## Upgrading
+
+It's necessary to specify the existing passwords while performing a upgrade to ensure the secrets are not updated with invalid randomly generated passwords. Remember to specify the existing values of the `rootUser.password`, `db.password` and `galera.mariabackup.password` parameters when upgrading the chart:
```
Comment (juan131) [inline bitnami/prometheus-operator/README.md]: Did you copy this from MariaDB Galera?

File: bitnami/prometheus-operator/templates/_helpers.tpl [hunk 14]
```diff
@@ -0,0 +1,377 @@
...
+{{ include "prometheus-operator.labels" . }}
+{{- end -}}
+
+{{/*
+Labels for alertmanager
+*/}}
+{{- define "prometheus-operator.alertmanager.labels" -}}
+app.kubernetes.io/name: {{ include "prometheus-operator.alertmanager.name" . }}
+{{ include "prometheus-operator.labels" . }}
+{{- end -}}
```
Comment (juan131) [inline bitnami/prometheus-operator/templates/_helpers.tpl]: I'm not very sure about the labels you chose... Shouldn't you use sth like this?
```
{{/*
Common Labels
*/}}
{{- define "prometheus-operator.labels" -}}
app.kubernetes.io/name: {{ include "prometheus-operator.name" . }}
helm.sh/chart: {{ include "prometheus-operator.chart" . }}
app.kubernetes.io/instance: {{ .Release.Name }}
app.kubernetes.io/managed-by: {{ .Release.Service }}
{{- if .Values.global.labels }}
{{ toYaml .Values.global.labels }}
{{- end }}
{{- end -}}

{{/*
Labels for operator
*/}}
{{- define "prometheus-operator.operator.labels" -}}
{{ include "prometheus-operator.labels" . }}
app.kubernetes.io/component: operator
{{- end -}}

{{/*
Labels for prometheus
*/}}
{{- define "prometheus-operator.prometheus.labels" -}}
{{ include "prometheus-operator.labels" . }}
app.kubernetes.io/component: prometheus
{{- end -}}

{{/*
Labels for alertmanager
*/}}
{{- define "prometheus-operator.alertmanager.labels" -}}
{{ include "prometheus-operator.labels" . }}
app.kubernetes.io/component: alertmanager
{{- end -}}
```

File: bitnami/prometheus-operator/templates/_helpers.tpl [hunk 15]
```diff
@@ -0,0 +1,377 @@
...
+{{ include "prometheus-operator.matchLabels" . }}
+{{- end -}}
+
+{{/*
+matchLabels for alertmanager
+*/}}
+{{- define "prometheus-operator.alertmanager.matchLabels" -}}
+app.kubernetes.io/name: {{ include "prometheus-operator.alertmanager.name" . }}
+{{ include "prometheus-operator.matchLabels" . }}
+{{- end -}}
```
Comment (juan131) [inline bitnami/prometheus-operator/templates/_helpers.tpl]: Same concern about labels.

File: bitnami/prometheus-operator/templates/_helpers.tpl [hunk 16]
```diff
@@ -0,0 +1,377 @@
...
+*/}}
+{{- define "prometheus-operator.validateValues" -}}
+{{- $messages := list -}}
+{{- $messages := without $messages "" -}}
+{{- $message := join "\n" $messages -}}
+
+{{- if $message -}}
+{{- printf "\nVALUES VALIDATION:\n%s" $message | fail -}}
+{{- end -}}
+{{- end -}}
```
Comment (juan131) [inline bitnami/prometheus-operator/templates/_helpers.tpl]: You don't have any validation. Is that expected?

File: bitnami/prometheus-operator/templates/alertmanager/pdb.yaml [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+apiVersion: policy/v1beta1
+kind: PodDisruptionBudget
+metadata:
+  name: {{ template "prometheus-operator.alertmanager.fullname" . }}
+  labels: {{- include "prometheus-operator.alertmanager.labels" . | nindent 4 }}
+spec:
+  selector:
+    matchLabels:
+      app: alertmanager
+      alertmanager: {{ template "prometheus-operator.alertmanager.fullname" . }}
```
Comment (juan131) [inline bitnami/prometheus-operator/templates/alertmanager/pdb.yaml]: Are these labels correct?

Comment (juan131) [reply bitnami/prometheus-operator/templates/alertmanager/pdb.yaml]: I guess it's the same you previously mentioned: https://github.com/coreos/prometheus-operator/blob/master/pkg/alertmanager/statefulset.go#L195-L197

File: bitnami/prometheus-operator/README.md [hunk 18]
```diff
@@ -0,0 +1,304 @@
...
+| `nameOverride`            | String to partially override `prometheus-operator.name` template with a string (will prepend the release name) | `nil`                                                   |
+| `fullnameOverride`        | String to fully override `prometheus-operator.fullname` template with a string                                 | `nil`                                                   |
+| `rbac.create`             | Wether to create & use RBAC resources or not                                                                   | `true`                                                  |
+| `rbac.apiVersion`         | Version of the RBAC API                                                                                        | `v1beta1`                                               |
+| `rbac.pspEnabled`         | PodSecurityPolicy                                                                                              | `true`                                                  |
```
[the remark this answers is not in the document]
Comment (sameersbn) [reply bitnami/prometheus-operator/README.md]: while they are not  under `.global` key, these configs actually have a global effect. I would not want to add more section to group `rbac` and the other configs. I would rather combine all the table back.

File: bitnami/prometheus-operator/README.md [hunk 19]
```diff
@@ -0,0 +1,304 @@
...
+| `alertmanager.additionalPeers`                  | AdditionalPeers allows injecting a set of additional Alertmanagers to peer with to form a highly available cluster | `[]`                                                                                                                                                                                                                                                |
+
+The above parameters map to the env variables defined in [bitnami/prometheus-operator](http://github.com/bitnami/bitnami-docker-prometheus-operator). For more information please refer to the [bitnami/prometheus-operator](http://github.com/bitnami/bitnami-docker-prometheus-operator) image documentation.
+
+Specify each parameter using the `--set key=value[,key=value]` argument to `helm install`. For example,
+
+```bash
+$ helm install --name my-release \
+  --set operator.logLevel=debug,
```
[the remark this answers is not in the document]
Comment (sameersbn) [reply bitnami/prometheus-operator/README.md]: good catch!

File: bitnami/prometheus-operator/README.md (hunk 13 above)
Comment (sameersbn) [reply bitnami/prometheus-operator/README.md]: yikes! :laughing:

File: bitnami/prometheus-operator/templates/_helpers.tpl (hunk 14 above)
Comment (sameersbn) [reply bitnami/prometheus-operator/templates/_helpers.tpl]: point taken

File: bitnami/prometheus-operator/templates/_helpers.tpl (hunk 16 above)
Comment (sameersbn) [reply bitnami/prometheus-operator/templates/_helpers.tpl]: yea. kept that around as a stub for future use

File: bitnami/prometheus-operator/templates/alertmanager/pdb.yaml (hunk 17 above)
Comment (sameersbn) [reply bitnami/prometheus-operator/templates/alertmanager/pdb.yaml]: yes. it's the same case as the previous comment

**Comment by sameersbn** [comment]
@javsalgar @juan131 i've made the requested changes, could you please take a look

**Pull request merged**